### DISCUSSION TITLE
What is the simplest way to *append* a middleware to `FastAPI.user_middleware`?

### DISCUSSION THREAD

**Discussion by znd4** [fastapi/fastapi#12478] [Questions]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from fastapi import FastAPI
import starlette.middleware
from opentelemetry.instrumentation.fastapi import FastAPIInstrumentor
from opentelemetry import trace

app = FastAPI()

def exception_log_middleware(_: fastapi.FastAPI):
    async def middleware(request: fastapi.Request, call_next):
        try:
            return await call_next(request)
        except Exception as e:
            print("exception caught")
            print(f"{trace.current_span()=}")
            raise

    return middleware


FastAPIInstrumentor.instrument_app(app)
app.add_middleware(exception_log_middleware)
# vs.
# app.user_middleware.append(
#     starlette.middleware.Middleware(
#         exception_log_middleware,
#     )
# )
```

### Description

I'm not sure how exactly it does this, but Opentelemetry's fastapi auto-instrumentation adds a middleware that adds an opentelemetry trace context to each request *before* my `app.add_middleware`s get run. This means that if I try to access the opentelemetry span / trace in *my* middleware, the opentelemetry instrumentation's middleware has already closed (because my middleware gets added outside of it).

I can get around this by `app.user_middleware.append`ing, as in the commented-out code above. Is there an easier way to do this without referring to `starlette.middleware.Middleware`? If not, would you be opent to a new `outer` boolean kwarg to `add_middleware` that defaults to `True` such that, `if not outer`, the new middleware will be `append`ed to `self.user_middleware` instead of the default "prepend" behavior 

### Operating System

macOS

### Operating System Details

n.a.

### FastAPI Version

0.109.2

### Pydantic Version

2.6.1

### Python Version

3.12

**Comment by YuriiMotov** [comment]
What if you just move `app.add_middleware(exception_log_middleware)` up?

```
app.add_middleware(exception_log_middleware)
FastAPIInstrumentor.instrument_app(app)
```

In FastAPI the order in which you call `app.add_middleware` matters

**Reply by znd4** [reply]
Oof, I spent a long time working on a more accurate reproducible example, but I think it was lost to a closed tab.

Suffice to say, I can't -- the `FastAPIInstrumentor` call is *actually* getting injected by [opentelemetry autoinstrumentation](https://opentelemetry.io/docs/zero-code/python/example/#execute-the-automatically-instrumented-server).

**Comment by estebanx64** [comment]
I don't know why you need to access to the span from opentelemetry, but there are some hooks that can help you on it.
https://opentelemetry-python-contrib.readthedocs.io/en/latest/instrumentation/fastapi/fastapi.html#request-response-hooks

Let me know if they are helpful, otherwise, can you explain a little bit more about what you need to do? Thanks in advance
On the other hand, using any ASGI middleware is legit too as we said here in the documentation, so the Starlette workaround can work too
https://fastapi.tiangolo.com/advanced/middleware/

FYI, we are using behind the scenes the user_middleware.insert with an index 0 from Starlette (Which makes me think the instrument from OpenTelemetry is waiting for the app to initialize everything including middleware and just adding itself on top of them, I'm not sure of it)

**Comment by YuriiMotov** [comment]
Looks like the author is not interested in this anymore. Let's close it for now to keep discussions cleaner.
If you managed to find the solution, please post it here.
If you still need help, please let us know
